### DISCUSSION TITLE
Multilanguage with Astroid

### DISCUSSION THREAD

**Discussion by iorbita** [templaza/astroid-framework#557] [Q&A] [answered]
Hi,
I thought that multilingual management with Astroid would make this tedious task easier and that there was no need to duplicate each template for each language, but apparently this is not the case according to the documentation I found on the [Joomlaplates website](https://www.joomlaplates.de/joomla-tips-und-tricks/multilanguage-mit-astroid.html)...

Since I have 8 different template styles for 3 languages, that makes 24 template styles...! This is annoying, it's a pity that no other solution is available to avoid copying all my template styles, or maybe I missed something? ...but I can't see how to assign a different main navigation menu and by language in the header options.

Too bad because it seems that this is the only element of the template that prevents keeping the same main menu for all languages.

Many thanks,
Lorenzo

![main-menus](https://github.com/templaza/astroid-framework/assets/566221/76710649-744f-4605-b597-87186c009a69)

![astroid_template_one-header](https://github.com/templaza/astroid-framework/assets/566221/a91c68d7-b884-4d38-87bd-a1c899f378e1)

**Comment by vtheod** [comment]
I double that. It would be perfect if main menu and mobile menu of astroid header were module positions. That way, building a multilingual site would be much easier.

**Comment by Chacapamac** [comment]
Do you know you have a special “**Astroid Menu Module**” — Don’t know if it can help You?

![Astroid Menu Module](https://github.com/templaza/astroid-framework/assets/1907276/326e6447-1e0b-4bfb-bb5d-d10824c3f4e7)

**Comment by iorbita** [comment]
Hi,
…indeed I missed that module, but finally the only difference between a native Joomla menu module and Astroid menu module is the breakpoint option. 
Anyway, it doesn't change much to the problem because in the template style what prevents having a common menu for all languages is the fact that you have to select the menu specific to the language and not a position or name module.

I don't know if this is a core Joomla restriction (I think so) and if other frameworks offer a solution to work around this limitation.

![astroid_template_one-header-menu](https://github.com/templaza/astroid-framework/assets/566221/21861874-f5db-4faa-b5b5-eb6f4c02bab6)

**Reply by rinka88** [reply]
Is it possible to skip the menu selection from the template, and it should select from the module position select the Module menu of the selected language

**Reply by rinka88** [reply]
In the Website Menu and Mobile menu section, make None. And it automatically selects the module assigned to this position

**Comment by Chacapamac** [comment]
> I don't know if this is a core Joomla restriction

Yes, at least, one Template layout by language and you need to point the correct language menu to each Layout styles. I’m not sure if you can bypass the way Astroid/Joomla work now.

**Comment by iorbita** [comment]
...just to understand multi-language with Astroid, okay to duplicate template styles which is still annoying and a pain to manage, but what about modules for each language?
Do you have to rename positions then reselect the module for each language in the template style? I hope not... but I can't see any other solution :(

I think another way of selecting modules (or better still, adding one more way) in the template style would be a new feature I'd like to see in Astroid, such as selecting the module by category, by name... or by language!

**Comment by Chacapamac** [comment, marked as answer]
@iorbita 

> okay to duplicate template styles which is still annoying and a pain to manage 

 —> All other Joomla frameworks work the same way, one Layout Style by language.

**About Modules**
If you show a module in a specific position and have 4 languages (and 4 Layout), you do not need to rename the position in the 4 layouts, you only need to specify the language of the 4 modules. Modules assigned to a language will not show on another language.

If you assigned the layout position to a specific module, you will need to assigned the specific module in each Layouts (language).

**Reply by rinka88** [reply]
Yootheme pro does not need to be duplicated. and in Cassiopeia too. I would also like this in Astroid

**Comment by iorbita** [comment]
@Chacapamac 
> About Modules
> If you show a module in a specific position and have 4 languages (and 4 Layout), you do not need to rename the position in the 4 layouts, you only need to specify the language of the 4 modules. Modules assigned to a language will not show on another language.
> 
> If you assigned the layout position to a specific module, you will need to assigned the specific module in each Layouts (language).


...you are right, I completely missed the multilingual options in modules ... so stupid from my part! 👎

**Reply by Chacapamac** [reply]
My father was saying “**Only the people that do nothing make no mistakes**.” 😉
**Happy to help!**

**Comment by rinka88** [comment]
But in general, is it possible to make a multilingual website without creating a template? like in the Cassiopeia template? otherwise in the Astroid template you need to duplicate styles for each language in order to assign a menu?

**Reply by Chacapamac** [reply]
**Two or more Template layouts for two or more languages assigned to two or more Main menus items** This is the Joomla way to create _(even in Cassiopea)_ multilingual website 

**I do not recommend** any Frameworks/Templates system that do not use All Joomla core systems and overrule or bypass Joomla Core abilities. Astroid is the perfect companion to Joomla because it only enhance the core abilities. No reinventing the wheel and complex hack of Joomla.

**Comment by rinka88** [comment]
@sonvnn I think we need to consider this decision again. Maybe add an Asteroid plugin to manage a multilingual website? When enabling multilingualism, add the site languages modules as the main menu. At least 3 languages for the site. To avoid duplicating styles for each language. And in the Mobile Menu section.

**Comment by iorbita** [comment]
Hi,
I agree that multilanguage in Joomla is a pain, you're quickly overwhelmed by an enormous number of menus if the site is complex, and updating isn't very resilient either, since with each new menu you have to duplicate the menu for each new language and associate it with the corresponding menu. In short, it's not my favourite feature in Joomla... and apparently that's how Joomla works.
It seems that Astroid also works like that, since navigation is entirely based on the menu, one menu per language, so one template per language...

So I'm thinking that every hack added to get round this problem is likely to cause serious incompatibility issues for every major Joomla update, if the developer suddenly disappears from the Joomla ecosystem (we've already been in this situation several times), on this point we are, unfortunately, deadlocked.

**Comment by Chacapamac** [comment]
I work and build ONLY multilingual websites since Joomfish on Joomla 1.5 and, for me, even on very complex websites, I find the native Joomla system fantastic and easy to work with, all that in the familiar Joomla environment.

**For me, this is probably the most important point in choosing Astroid in the first place**, because it completely relies on the core Joomla system. Doing otherwise will make me look to other frameworks.

Sadly these other Frameworks, make the mistake of trying to reinvent the wheel inside Joomla and have a tendency to make you dependent on their system, and that is a big no, no for somebody building a website that endures time. 

My advice is to work with Joomla Multilingual for a while. I promise it will soon be second nature, and you will discover the power behind it.

**Reply by colognella** [reply]
Just asking, as I stumbled upon this thread with exact the same question: My main problem (if I do get it right) with this approach is the fact that I don't need different layouts for different languages, but the opposite: every single change I need to do on the design and layout (in the process of developing and later) I need to do again for every single language. CSS-fixes, font size corrections, slight color changes, background images for sections, everything. (OK, custom css could point to the same external file for all styles..)  The duplicate template style wouldn't be a kind of "child" style that inherits all the design stuff, would it?

**Reply by rinka88** [reply]
If there is duplication of styles, then yes. Each style has its own settings.

**Reply by colognella** [reply]
And this redundant way to work is the preferred Joomla way of doing multilanguage right? Just to make sure … (yes, I read the thread, however, I'm a bit flabbergasted)

**Reply by rinka88** [reply]
this is inherent in almost all club templates. (Duplicate styles)

**Comment by rinka88** [comment]
@sonvnn maybe it’s worth adding to the section Site Menu and Mobile menu, position selection? then the modules of this position will be displayed. And add a menu module to this position

**Reply by rinka88** [reply]
If this is implemented (output through a module), then it will be a huge plus for a multilingual site; there will be no need to duplicate the template style for the sake of the site language. Now this is not possible due to the rigid assignment of the menu, without choice (there is a choice, but only one)

**Comment by Chacapamac** [comment]
` there will be no need to duplicate the template style for the sake of the site language.` **This is how Joomla Work.**

**Reply by rinka88** [reply]
That's why they output everything through modules. since the content language can be assigned to modules.

**Reply by rinka88** [reply]
![303365295-21861874-f5db-4faa-b5b5-eb6f4c02bab6](https://github.com/templaza/astroid-framework/assets/40716604/c99d345c-1581-4f03-badf-6e1644928938)
if in these places there were to select the positions of the modules. that would be great. You assign the module menu to this position, and you don’t have to duplicate the style.

**Reply by rinka88** [reply]
@sonvnn I think, pay attention to this and implement what users ask for )

**Reply by sonvnn** [reply]
@rinka88 @Chacapamac  I'm always listening to your problems and following your conversations. Improving Menu display using Modules is a good idea and Joomla is doing it.

But I'm weighing the risks of making this improvement. You should know that the current Menu display has been around since Joomdev started with Astroid. And when I took over from Joomdev I had no choice but to continue inheriting it. As far as I know, if we change the Menu display to a different approach. Current websites, old templates currently using Astroid will have the issue that the Menu will disappear on their website when they update this version, immediately. It will create a wave of complaints from many users. 

Because to display the menu, they will have to go to the Module Manager and create a Menu Module then configure its position so that it displays again on their site. And that will definitely make them angry when the website that is operating normally is affected.

I think you understand the point. We should think about coming up with a solution that is safe and has the least impact on old Astroid users.

**Reply by rinka88** [reply]
I think this needs to be implemented for version 3.1.0 and fully describe the transition in documentation. Since 3.1.0 there should be new improvements with changes during the transition, which must be read in the release description. (If possible, then leave the current state and add the ability to use new improvements)

**Comment by Chacapamac** [comment]
**My only advice here is to Astroid to work in TOTAL tandem with Joomla** and not become one of these ‘reinvent the wheel” framework.

**Reply by rinka88** [reply]
Displaying the position of modules This is the basics of joomla. Astroid doesn't break anything.
Any mod in joomla is displayed through a position in the template.
(And if you rigidly define one module without selecting it, this makes the work difficult)
In the Cassiopeia template, the Menu is displayed through the module position, and without duplicating the template style for each language. But it only duplicates menu modules for each language
https://www.youtube.com/watch?v=nKR0OL2fOmg

**Comment by Chacapamac** [comment]
Ok first, I will put a Chat GPT answer here to the question “In Joomla, do you need another template style for each language?”

**Response:**

In Joomla, you do not necessarily need a separate template style for each language, but it is a common practice to create a template style for each language when setting up a multilingual website. This allows you to manage different layouts, module assignments, and other style-related settings for each language individually.

Here's a brief overview of the steps involved in setting up a multilingual site in Joomla:

1. **Install Languages**: Install the languages you want to use on your Joomla site.
2. **Enable Language Plugins**: Enable the necessary language plugins to manage content associations and filtering based on language selection.
3. **Add Multilingual Content**: Create categories, articles, and modules for each language.
4. **Set Up Multilingual Menus**: Create menu items for each language and set default home pages for each language.
5. **Create Template Styles for Each Language**: Navigate to `System > Site Template Styles`, select your default template style, and duplicate it for each language¹.
6. **Assign Template Styles**: Assign the duplicated template styles to the respective language's content.
7. **Add a Language Switcher**: Implement a language switcher module to allow users to switch between different languages.

By following these steps, you can effectively manage a multilingual Joomla site with different template styles for each language, ensuring that each language's content is presented with the appropriate layout and design¹.

Source: Conversation with Copilot, 2024-05-22
(1) How to create a multilingual website in Joomla 4 - step by step .... https://www.joomlart.com/blog/how-to-create-a-multilingual-website-in-joomla-4-step-by-step-tutorial.
(2) [SOLVED] multi-language site, copy template style for each language .... https://www.joomlaplates.com/forum/joomla-4-templates/11111-solved-multi-language-site,-copy-template-style-for-each-language.html.
(3) How to create a multilingual website in Joomla 4 - comprehensive steps .... https://www.cmsjunkie.com/blog/post/how-to-create-a-multilingual-website-in-joomla-4-comprehensive-steps.
(4) Can I have different templates for different languages in a Joomla .... https://stackoverflow.com/questions/18133751/can-i-have-different-templates-for-different-languages-in-a-joomla-multilanguage.

**Reply by rinka88** [reply]
All the links you provided were for club templates.
Even the second link describes the Astroid problem that was discussed in this thread.
Even he is compared to Cassiopeia, with the default template of Joomla.
This post also contains all the actions by default
https://kevinsguides.com/guides/webdev/joomla/content/multilingual

**Comment by Chacapamac** [comment]
As @sonvnn  say we need to think about this....  

**_Some ideas and questions that come to my mind as a BRAINSTORMING ⚡️🧠⚡️_**

1. Can Astroid already do that with the “Astroid Menu Module” or it will need a switch to pass from the original one layout by language to a single layout with one “Astroid Menu Module” by language?  
2. Are the Astroid “Header” different positioning abilities be use in an “Astroid Menu Module”?
3. The logo is sometimes different in different language (to keep in mind....)
4. We NEED the keep the ability to have a different Layout style by language.
5. Backward compatibility is an issue what about a switch in the Layout Manager to choose between the default original “one layout by language” to a single layout with one “Astroid Menu Module” by language? 

Will think to something else later.🙄

**Reply by rinka88** [reply]
Yes, I agree that when you need to apply your own template style and logos for each language, you need to duplicate the style.
But there are sites that don’t need different styles; you don’t want to duplicate templates.
And it happens that you need to change some color of blocks or logos, or add blocks, then you need to repeat each action in each style. What if there are several such sites?
This is why I say that it is advisable to add duplication without the possibility if all language styles are the same.
And the fact that Backward Compatibility means that sometimes you have to apply an improvement in order to move forward.

**Reply by rinka88** [reply]
@colognella here and for this reason I wrote that for those who do not need duplication of styles, you need one template. If there is no need to change for each language

**Comment by Chacapamac** [comment]
Just thinking! 🤔
💡IDEAS
— Yes, it will be great in Astroid if we have both choice to create a multilingual website.
**1.** The Joomla way, One layout style by language
_**NOTE:** In Astroid, after any Layout style change, it take seconds to export/import ALL settings from a Layout to another. In a Multilingual perspective, You simply need to reset the target language menu._

**2.** An hypothetical second way with a SINGLE Astroid layout that can be switched to do...
— target each languages menus in the same layout

@sonvnn 

> One thing is sure. The “Joomla way” have to be the “Default”. — Maybe a switch in the layout to turn a “Normal” layout into a Multilingual “Layout” ????????

**Comment by sonvnn** [comment]
Hi @iorbita @vtheod @rinka88 @Chacapamac @colognella 

Finally, I found a way to add module position to Astroid Header here https://github.com/templaza/astroid-framework/commit/afef57c80ea922e59cd6cf55b72189290bb2adb5

### To test this function:

1. Download v3-branch and upload to your test site.
2. Module manager -> Create a Astroid Menu module -> assign to a position
3. Go to your template -> Template Options -> Header section 
4. Find "Site Menu Method" and select "Module Position" which contains Astroid Menu module. You can do the same task with Mobile Menu

![Screen Shot 2024-05-29 at 15 48 59](https://github.com/templaza/astroid-framework/assets/3204103/cd840193-5bb7-4204-9dfb-5d3bf8459f8b)

Go to front end and check it.

Thanks & Best Regards,
Sonny

**Reply by rinka88** [reply]
This is one of the most excellent news.

**Reply by colognella** [reply]
@sonvnn some feedback after my first test – besides my deepest admiration and gratefulness …. Unfortunately, I don't have a "vanilla" local testing instance at the moment, only a project instance with a Joomlaplates template and 2 languages. If somebody would try to reproduce this phenomenon, I'd greatly appreciate: 

- do everything you wrote above
- create 2 (or more) menu modules in 2 languages (languages explicitely set for each module)
- assign both to the menu position that is set in Astroid -> the menus are (visually) displayed beautifully as expected. 
- Check the developer console or the source code: in my instance, there are 2 DOM sub trees with menus, the one in the active language works as expected, and an empty one:

<img width="703" alt="Bildschirmfoto 2024-05-29 um 18 58 40" src="https://github.com/templaza/astroid-framework/assets/26031183/b3f7848f-7060-481a-a973-efcd86963baa">

Not sure if this is the expected behaviour, it seems the "other" language's menu module is just emptied, not completely removed from the DOM tree. I noticed it only because I append something to the menu with jQuery, and it was duplicated (which could be fixed, however you might want to know)

**Reply by sonvnn** [reply]
Which type of header are you using? If you can please send me your test url here. I will check it.

Best regards,
Sonny

**Reply by rinka88** [reply]
@colognella Could this be a problem with your club template? Try it on a clean Astroid, I didn’t find this problem there

**Reply by colognella** [reply]
ok, thanks for your kind support, it seems it's a different issue on my instance, that displays ANY (other) language module without filtering out the language. I need to dig into this deeper, however, it's not an issue with your Astroid branch (UPDATE: Advanced Module Manager Plugin apparently is the culprit). Thanks again! Looking forward to the release!

**Comment by colognella** [comment]
@sonvnn just one more maybe dumb question regarding this highly appreciated feature: in your screenshot you assign an **astroid-header-mobilemenu** position for the mobile menu. Is this something the template would need to offer (or I would need to create) or something which would be inherent to astroid? Asking for my project which doesn't seem to have this position (nor another mobile menu position)
Thx

**Reply by rinka88** [reply]
these are the positions of template modules (zero, one, two), it will probably be released tomorrow. For now, you can write it manually directly in the module settings.

**Comment by rinka88** [comment]
@sonvnn It seems it's time to update the templates too. since there is no position initially, the menus are duplicated

**Reply by rinka88** [reply]
Don't forget to look at this error. (Menu created using Astroid menu and displayed via position)
